compiler/rustc_builtin_macros/src/format.rs

```diff
@@ -765,20 +764,26 @@ impl<'a, 'b> Context<'a, 'b> {
         // format! string are shoved into locals. Furthermore, we shove the address
         // of each variable because we don't want to move out of the arguments
         // passed to this function.
-        for (i, e) in self.args.into_iter().enumerate() {
-            for arg_ty in self.arg_unique_types[i].iter() {
-                args.push(Context::format_arg(self.ecx, self.macsp, e.span, arg_ty, i));
-            }
-            heads.push(self.ecx.expr_addr_of(e.span, e));
-        }
-        for pos in self.count_args {
+        let heads: Vec<_> =
+            self.args
+                .into_iter()
+                .enumerate()
+                .map(|(i, e)| {
+                    args.extend(self.arg_unique_types[i].iter().map(|arg_ty| {
```

Comment:
Mutating captured variables in an iterator adapter seems pretty iffy.

Reply:
It is, here the situation is quite complex indeed, with nested iterators.

Since this was mostly motivated by performance and it doesn't seem that it was improved in any way (and some of the changes maybe worsen readability rather than improve it), do we really want to merge this?

Reply:
I do find the other changes more readable. I think there may be a way to write this particular section in a nice way too (`flat_map` maybe?), but I think it is probably better to just revert this specific change and consider changes here as a separate PR.